Fixed NSInternalInconsistencyException & Swift runtime failure: arithmetic overflow

azisramdhan opened the pull request:
### Litewallet internal issues

[13](https://github.com/litecoin-foundation/litewallet-engineering/issues/13)
[16](https://github.com/litecoin-foundation/litewallet-engineering/issues/16)

### Xcode Organizer
<img width="1680" alt="Screenshot 2024-12-22 at 08 46 23" src="https://github.com/user-attachments/assets/1560c675-f5f9-498c-a2ab-b8584a577817" />
<img width="1680" alt="Screenshot 2024-12-22 at 08 45 39" src="https://github.com/user-attachments/assets/08c84bd8-b10e-432c-a1be-6ead0796be82" />

azisramdhan committed:
Fixed NSInternalInconsistencyException

azisramdhan committed:
Handle Swift runtime failure: arithmetic overflow

hassanashraf92 commented on litewallet/Platform/Extensions.swift:
```diff
@@ -558,3 +558,17 @@ extension Dictionary where Key: ExpressibleByStringLiteral, Value: Any {
 	}
 }
+
+extension UInt64 {
```
It's recommended to move the UInt64 extension to a dedicated file named something like `UInt64+SafeArithmetic.swift` or `UInt64+Extension.swift`.

hassanashraf92 replied:
Consider Extending a Protocol (Optional): If similar safe arithmetic methods are needed for other numeric types (e.g., Int, Double), you could introduce a protocol for safe arithmetic. This promotes code reuse and reduces duplication.
```
protocol SafeArithmetic {
    func safeAddition(_ value: Self) -> Self?
    func safeSubtraction(_ value: Self) -> Self?
}
```

hassanashraf92 commented on litewallet/Platform/Extensions.swift:
```diff
@@ -558,3 +558,17 @@ extension Dictionary where Key: ExpressibleByStringLiteral, Value: Any {
 }
+
+extension UInt64 {
+	/// Performs a safe addition, returning `nil` if overflow occurs.
```
While the existing documentation is clear, it can include more details about when these methods are useful.
/// Adds two `UInt64` values safely, returning `nil` if overflow occurs.
/// - Parameter value: The value to add.
/// - Returns: The sum of the two values or `nil` if the operation overflows.

hassanashraf92 commented on litewallet/Platform/Extensions.swift:
```diff
@@ -558,3 +558,17 @@ extension Dictionary where Key: ExpressibleByStringLiteral, Value: Any {
+		return overflow ? nil : result
+	}
+
+	/// Performs a safe subtraction, returning `nil` if underflow occurs.
```
Same comment for documentation above.

hassanashraf92 commented on litewallet/ViewModels/Transaction.swift:
```diff
@@ -36,18 +36,29 @@ class Transaction {
+        
+        // Safely calculate fee = fee + opsAmount
+		guard let safeFee = fee.safeAddition(opsAmount) else { return nil }
+		self.fee = safeFee
```
Seems that the same value (fee) is used multiple times across the class functions. We can cache fee in a local variable if it's not modified between operations.

```
guard let safeFee = fee.safeAddition(opsAmount) else { return nil }
self.fee = safeFee

let cachedFee = safeFee // Cache for reuse
```

hassanashraf92 commented on litewallet/ViewModels/Transaction.swift:
```diff
@@ -36,18 +36,29 @@ class Transaction {
-
-		self.fee = fee + opsAmount
+        
+        // Safely calculate fee = fee + opsAmount
```
While comments like `// Safely calculate fee = fee + opsAmount` provide context, they are overly verbose and repeat what the code already expresses.
For example; `// Ensure the new fee calculation doesn't overflow`

hassanashraf92 commented on litewallet/ViewModels/Transaction.swift:
```diff
@@ -36,18 +36,29 @@ class Transaction {
-		self.fee = fee + opsAmount
+        
+        // Safely calculate fee = fee + opsAmount
+		guard let safeFee = fee.safeAddition(opsAmount) else { return nil }
```
Some variable names (e.g., safeSum, safeFee) are generic and don't convey their specific role in the calculation. Use more descriptive names for better understanding.
```
guard let totalReceivedAndFee = amountReceived.safeAddition(fee) else { return nil }
```

hassanashraf92 commented on litewallet/ViewModels/Transaction.swift:
```diff
@@ -36,18 +36,29 @@ class Transaction {
-		let amountSent = wallet.amountSentByTx(tx) - opsAmount
 
-		if amountSent > 0, (amountReceived + fee) == amountSent {
+		// Safely calculate amountSent = wallet.amountSentByTx(tx) - opsAmount
```
Same comment for documentation.

hassanashraf92 commented on litewallet/ViewModels/Transaction.swift:
```diff
@@ -36,18 +36,29 @@ class Transaction {
+			satoshis = safeAmountSent
+		} else if safeAmountSent > 0 {
+			// Safely calculate (safeAmountSent - amountReceived - fee)
+			guard let safeSatoshis = safeAmountSent
```
Multi-line guard statements can be difficult to read and debug. Break these into smaller, intermediate steps for clarity.

```
guard let intermediateSatoshis = safeAmountSent.safeSubtraction(amountReceived),
      let finalSatoshis = intermediateSatoshis.safeSubtraction(fee) else { return nil }
direction = .sent
satoshis = finalSatoshis
```

hassanashraf92 commented on litewallet/Views/DefaultCurrencyViewController.swift:
```diff
@@ -22,11 +22,14 @@ class DefaultCurrencyViewController: UITableViewController, Subscriber {
 		didSet {
```
The didSet block contains too much logic. It's better to extract this logic into helper functions to make the didSet block cleaner and more focused.

```
updateCurrencyRows(oldCurrencyCode: oldValue)
```

(Optional but Recommended): Avoid UI Updates in `didSet`

hassanashraf92 commented on litewallet/Views/DefaultCurrencyViewController.swift:
```diff
@@ -22,11 +22,14 @@ class DefaultCurrencyViewController: UITableViewController, Subscriber {
-			setExchangeRateLabel()
+            
+            Task { @MainActor in
+                tableView.beginUpdates()
```
(Optional): Use `performBatchUpdates` for Table View Changes.

hassanashraf92 reviewed (changes requested):
General Comment: The indentation in the provided code appears inconsistent with standard Swift style guidelines.
Make sure to use `control + i` to fix any inconsistent indentation.

azisramdhan committed:
Enhance documentation for safe addition and subtraction methods

- Added detailed explanations of use cases for safeAddition and safeSubtraction methods in UInt64 extension.
- Clarified the behavior of these methods when handling overflow and underflow.
- Highlighted their importance in contexts requiring safe and predictable arithmetic operations.

azisramdhan committed:
Reformat code using Xcode's indentation

- Applied consistent indentation using Xcode's Ctrl + I shortcut to improve code readability.
- No functional changes were made.

azisramdhan committed:
Refactor fee and transaction amount calculations to ensure safety and clarity

- Cache the total fee in a local variable for reuse.
- Safely calculate the amount sent by subtracting opsAmount to prevent underflow.
- Ensure the total received and fee are within bounds during the calculation.
- Simplify the process of determining transaction direction (moved, sent, received) by using safe arithmetic.

azisramdhan committed:
Refactor defaultCurrencyCode didSet logic into helper function

- Moved the logic for updating currency rows into a separate helper function () to improve readability and maintainability.
- Cleaned up the didSet block to focus on calling the helper function.

azisramdhan commented:
Thank you, @hassanashraf92, for the valuable feedback. I have updated the code based on your suggestions.

hassanashraf92 reviewed (approved):
Great Job!
